### DISCUSSION TITLE
Two tables on the same page are extracted as one

### DISCUSSION THREAD

**Discussion by filips123** [jsvine/pdfplumber#872] [Ask for help with specific PDFs]
## Describe the bug

On a PDF document where there are multiple tables on the same page, both of them are extracted as one table. This causes misaligned and empty columns.

## Code to reproduce the problem

```py
import pdfplumber
import pandas as pd

file = pdfplumber.open("example document.pdf")

page = file.pages[0]
tables = page.extract_tables()
df = pd.DataFrame(tables[0][1:], columns=table[0][0])
```

## PDF file

[example document.pdf](https://github.com/jsvine/pdfplumber/files/11303433/example.document.pdf)

This is an example document with redacted information. Real documents have a similar structure, but with more pages, sometimes with multiple and different tables.

## Expected behavior

There should be two tables:

* One with columns: ODSOTNI UČITELJ/ICA, URA, RAZRED, UČILNICA, NADOMEŠČA, PREDMET, OPOMBA
* Another with columns: RAZRED, URA, UČITELJ/ICA, PREDMETA, UČILNICA, OPOMBA

Text between those two tables (MENJAVA UR) should not be in any of tables.

## Actual behavior

*What actually happened, instead?*

Tables are extracted as one. This causes a lot of extra and misaligned columns with `None`:

![actual table](https://user-images.githubusercontent.com/16626308/233834929-27ece6f8-0ad8-4557-9f5f-a5205ca2fd74.png)

## Screenshots

![table finder debug](https://user-images.githubusercontent.com/16626308/233834966-ebef7f91-c822-4938-8d4d-8a47313c1cc2.png)

## Environment

- pdfplumber version: 0.9.0
- Python version: 3.10
- OS: Windows

## Additional context

PDFs that I need to parse all share a similar layout with around 5 different table formats. However, not all files use all of those table formats, and not all tables have the same height, making it impossible to just crop the page before extracting tables.

Maybe this can be solved by adding an option to only extract tables where there are visible lines around cells?

**Comment by jsvine** [comment]
Hi @filips123, and thanks for your interest in this library.  

> Maybe this can be solved by adding an option to only extract tables where there are visible lines around cells?

Unfortunately, there's no foolproof way (without adding a ton of post-processing) of identifying what lines are "visible" vs. not in a PDF, so `pdfplumber` chooses to take the conservative route and does not try to guess that. But if you're working with a relatively consistent set of PDFs, you may be able to achieve your goals via the `Page.filter` method, filtering out lines that meet certain criteria. See, e.g., here (and elsewhere in the discussions forum): https://github.com/jsvine/pdfplumber/issues/311#issuecomment-726958591

In the case of this specific example, something quite close to that comment above does the trick, just tweaking this function:

```python
def keep_visible_lines(obj):
    if obj['object_type'] == 'rect':
        # Slightly tweaked here, because this PDF uses RGB values instead of grayscale values
        return obj['non_stroking_color'] == [0, 0, 0]
    return True
```

**Reply by filips123** [reply]
Thank you, this works fine!
